numba/targets/linalg.py

```diff
@@ -497,223 +504,110 @@ def dot_3(context, builder, sig, args):
         else:
             assert 0
 
+fatal_error_sig = types.intc()
+fatal_error_func = types.ExternalFunction("numba_fatal_error", fatal_error_sig)
 
-def call_xxgetrf(context, builder, a_type, a_shapes, a_data, ipiv, info):
-    """
-    Call the LAPACK gettrf function for the given argument.
 
-    This function computes the LU decomposition of a matrix.
-    """
-    # XXX for ipiv, we are assuming a Fortran (LAPACK) int is the same size
-    # as a C int.
-    fnty = ir.FunctionType(ll_intc,
-                           [ll_char,                       # kind
-                            intp_t, intp_t,                # m, n
-                            ll_void_p, intp_t,             # a, lda
-                            ll_intc_p, ll_intp_p,          # ipiv, info
-                            ])
+@jit(nopython=True)
+def _check_finite_matrix(a):
+    for v in np.nditer(a):
+        if not np.isfinite(v.item()):
+            raise np.linalg.LinAlgError(
+                "Array must not contain infs or NaNs.")
 
-    fn = builder.module.get_or_insert_function(fnty, name="numba_xxgetrf")
 
-    kind = get_blas_kind(a_type.dtype)
-    kind_val = ir.Constant(ll_char, ord(kind))
+def _check_linalg_matrix(a, func_name):
+    if not isinstance(a, types.Array):
+        raise TypingError("np.linalg.%s() only supported for array types"
+                          % func_name)
+    if not a.ndim == 2:
+        raise TypingError("np.linalg.%s() only supported on 2-D arrays."
+                          % func_name)
+    if not isinstance(a.dtype, (types.Float, types.Complex)):
+        raise TypingError("np.linalg.%s() only supported on "
+                          "float and complex arrays." % func_name)
 
-    if a_type.layout == 'F':
-        m, n = a_shapes
-        lda = a_shapes[0]
-    else:
-        n, m = a_shapes
-        lda = a_shapes[1]
 
-    res = builder.call(fn, (kind_val, m, n,
-                            builder.bitcast(a_data, ll_void_p), lda,
-                            ipiv, info
-                            ))
-    check_lapack_return(context, builder, res)
+@jit(nopython=True)
+def _inv_err_handler(r):
+    if r != 0:
+        if r < 0:
+            fatal_error_func()
+            assert 0   # unreachable
+        if r > 0:  # this condition should be caught already above!
```

Comment:
I think its an artefact of moving this error handler out. It would have been in the case of calling `xxgetrf` then `xxgetri`, in which if the matrix were singular it'd have been caught by the former making the check on the latter redundant. Net result, I'll delete the comment.